system-contracts/contracts/libraries/Utils.sol

```diff
@@ -106,8 +129,74 @@ library Utils {
             EfficientCall.sha(_bytecode) &
             0x00000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
         // Setting the version of the hash
-        hashedBytecode = (hashedBytecode | bytes32(uint256(1 << 248)));
+        hashedBytecode = (hashedBytecode | bytes32(uint256(ERA_VM_BYTECODE_FLAG) << 248));
         // Setting the length
         hashedBytecode = hashedBytecode | bytes32(lengthInWords << 224);
     }
+
+    // the real max supported number is 2^16, but we'll stick to evm convention
+    uint256 internal constant MAX_EVM_BYTECODE_LENGTH = (2 ** 16) - 1;
+
+    /// @notice Validate the bytecode format and calculate its hash.
+    /// @param _bytecode The EVM bytecode to hash.
+    /// @return hashedEVMBytecode The 32-byte hash of the EVM bytecode.
+    /// Note: The function reverts the execution if the bytecode has non expected format:
+    /// - Bytecode bytes length is not a multiple of 32
+    /// - Bytecode bytes length is greater than 2^16 - 1 bytes
+    /// - Bytecode words length is not odd
+    function hashEVMBytecode(bytes calldata _bytecode) internal view returns (bytes32 hashedEVMBytecode) {
+        // Note that the length of the bytecode must be provided in 32-byte words.
+        if (_bytecode.length % 32 != 0) {
+            revert MalformedBytecode(BytecodeError.Length);
+        }
+
+        if (_bytecode.length > MAX_EVM_BYTECODE_LENGTH) {
+            revert MalformedBytecode(BytecodeError.EvmBytecodeLength);
+        }
+
+        uint256 lengthInWords = _bytecode.length / 32;
+        // bytecode length in words must be odd
+        if (lengthInWords % 2 == 0) {
+            revert MalformedBytecode(BytecodeError.WordsMustBeOdd);
+        }
+
+        hashedEVMBytecode =
+            EfficientCall.sha(_bytecode) &
+            0x00000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+
+        // Setting the version of the hash
+        hashedEVMBytecode = (hashedEVMBytecode | bytes32(uint256(EVM_BYTECODE_FLAG) << 248));
+        hashedEVMBytecode = hashedEVMBytecode | bytes32(_bytecode.length << 224);
```

Comment:
Should it be ?
```suggestion
        hashedEVMBytecode = hashedEVMBytecode | bytes32(lengthInWords << 224);
```

Reply:
Actually it should be length of unpadded EVM bytecode in bytes. This will make `EXTCODESIZE` implementation in Emulator a lot simpler. Right now this encoding in versioned bytecode hash (_padded_ evm code length) actually doesn't make any sense

Reply:
WIP: https://github.com/matter-labs/era-contracts/pull/1066